File: RepoTransBench/target_projects/Python/Go/tusen-ai_MV2D/public_tests/public_configs_exp_attributes_test.go
package public_tests

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/tusanai/mv2d/tests"
)

var publicExpConfigs = []string{
	"configs/mv2d/exp/mv2d_r50_frcnn_single_frame_roi_1408x512_ep24.py",
	"configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep24.py",
	"configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep72.py",
}

func TestPublicModelConfigKeys(t *testing.T) {
	for _, configPath := range publicExpConfigs {
		module, err := tests.LoadPythonModule(configPath)
		assert.NoError(t, err)
		model := module["model"].(map[string]interface{})
		requiredKeys := []string{"type", "use_grid_mask", "base_detector", "neck", "roi_head", "train_cfg"}
		for _, k := range requiredKeys {
			_, ok := model[k]
			assert.Truef(t, ok, "Missing key: %s in %s", k, configPath)
		}
		gm := model["use_grid_mask"].(map[string]interface{})
		found := false
		for _, v := range gm {
			b, isBool := v.(bool)
			f, isFloat := v.(float64)
			if isBool && b {
				found = true
			}
			if isFloat {
				found = true
			}
		}
		assert.True(t, found, "At least one grid mask option should be True or a float")
	}
}

func TestPublicROIHeadAndNested(t *testing.T) {
	configPath := "configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep24.py"
	module, err := tests.LoadPythonModule(configPath)
	assert.NoError(t, err)
	model := module["model"].(map[string]interface{})
	rh := model["roi_head"].(map[string]interface{})
	_, ok := rh["bbox_roi_extractor"]
	assert.True(t, ok)
	_, ok = rh["bbox_head"]
	assert.True(t, ok)
	_, ok = rh["query_generator"]
	assert.True(t, ok)
	_, ok = rh["pe"]
	assert.True(t, ok)
	if bhRaw, ok := rh["bbox_head"]; ok {
		bh := bhRaw.(map[string]interface{})
		_, ok := bh["transformer"]
		assert.True(t, ok)
		_, ok = bh["bbox_coder"]
		assert.True(t, ok)
		numClasses, hasNum := bh["num_classes"]
		if hasNum {
			switch nc := numClasses.(type) {
			case int:
				assert.True(t, nc > 0)
			case float64:
				assert.True(t, nc > 0)
			}
		}
		if lossCls, ok := bh["loss_cls"]; ok {
			_, isDict := lossCls.(map[string]interface{})
			assert.True(t, isDict)
		}
	}
	if peRaw, ok := rh["pe"]; ok {
		pe := peRaw.(map[string]interface{})
		_, ok := pe["positional_encoding"]
		assert.True(t, ok)
	}
}

func TestPublicTrainCfgDetections(t *testing.T) {
	for _, configPath := range publicExpConfigs {
		module, err := tests.LoadPythonModule(configPath)
		assert.NoError(t, err)
		model := module["model"].(map[string]interface{})
		trainCfg := model["train_cfg"].(map[string]interface{})
		if detRaw, ok := trainCfg["detection_proposal"]; ok {
			det := detRaw.(map[string]interface{})
			_, ok := det["score_thr"]
			assert.True(t, ok)
			if nmsRaw, ok := det["nms"]; ok {
				nms := nmsRaw.(map[string]interface{})
				_, okIou := nms["iou_threshold"]
				_, okCls := nms["class_agnostic"]
				assert.True(t, okIou || okCls)
			}
		}
	}
}

func TestPublicCodeWeightsVariations(t *testing.T) {
	configPaths := []string{
		"configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep24.py",
		"configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep72.py",
	}
	for _, configPath := range configPaths {
		module, err := tests.LoadPythonModule(configPath)
		assert.NoError(t, err)
		bboxHead := module["model"].(map[string]interface{})["roi_head"].(map[string]interface{})["bbox_head"].(map[string]interface{})
		codeWeights := bboxHead["code_weights"].([]interface{})
		assert.Len(t, codeWeights, 10)
		allOne := true
		for _, w := range codeWeights {
			_, isNum := w.(float64)
			assert.True(t, isNum)
			if isNum && w.(float64) != 1.0 {
				allOne = false
			}
		}
		assert.False(t, allOne, "Not all code_weights should be exactly 1.0")
	}
}

func TestPublicOperationOrderAndCP(t *testing.T) {
	for _, configPath := range publicExpConfigs {
		module, err := tests.LoadPythonModule(configPath)
		assert.NoError(t, err)
		bboxHead := module["model"].(map[string]interface{})["roi_head"].(map[string]interface{})["bbox_head"].(map[string]interface{})
		decoderRaw, hasDecoder := bboxHead["transformer"].(map[string]interface{})["decoder"]
		if hasDecoder {
			decoder := decoderRaw.(map[string]interface{})
			tlayersRaw, ok := decoder["transformerlayers"]
			if ok {
				tlayers, _ := tlayersRaw.(map[string]interface{})
				oo, ok2 := tlayers["operation_order"]
				assert.True(t, ok2)
				switch oo.(type) {
				case []interface{}:
				default:
					assert.Fail(t, "operation_order must be list/tuple")
				}
				withCpSet, cpOk := tlayers["with_cp"]
				assert.True(t, cpOk)
				b, isBool := withCpSet.(bool)
				assert.True(t, isBool)
				assert.True(t, b || !b)
			}
		}
	}
}

File: RepoTransBench/target_projects/Python/Java/tusen-ai_MV2D/src/test/java/com/tusenai/mv2d/publictests/PublicConfigsExpAttributesTest.java
package com.tusenai.mv2d.publictests;

import com.tusenai.mv2d.ConfigLoader;
import com.google.gson.*;
import org.junit.jupiter.api.*;

import static org.junit.jupiter.api.Assertions.*;

public class PublicConfigsExpAttributesTest {

    private static final String[] PUBLIC_EXP_CONFIGS = {
        "configs/mv2d/exp/mv2d_r50_frcnn_single_frame_roi_1408x512_ep24.json",
        "configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep24.json",
        "configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep72.json"
    };

    @Test
    public void testPublicModelConfigKeys() throws Exception {
        for (String configPath : PUBLIC_EXP_CONFIGS) {
            JsonObject module = ConfigLoader.loadConfig(configPath).getAsJsonObject("module");
            String[] requiredKeys = {"type", "use_grid_mask", "base_detector", "neck", "roi_head", "train_cfg"};
            for (String key : requiredKeys) {
                assertTrue(module.has(key), "Missing key: " + key + " in " + configPath);
            }
            JsonObject gm = module.getAsJsonObject("use_grid_mask");
            boolean anyTrueOrFloat = false;
            for (String field : gm.keySet()) {
                JsonElement val = gm.get(field);
                if (val.isJsonPrimitive()) {
                    if (val.getAsJsonPrimitive().isBoolean() && val.getAsBoolean())
                        anyTrueOrFloat = true;
                    if (val.getAsJsonPrimitive().isNumber())
                        anyTrueOrFloat = true;
                }
            }
            assertTrue(anyTrueOrFloat, "At least one grid mask option should be True or a float");
        }
    }

    @Test
    public void testPublicRoiHeadAndNested() throws Exception {
        String configPath = "configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep24.json";
        JsonObject module = ConfigLoader.loadConfig(configPath).getAsJsonObject("module");
        assertTrue(module.has("roi_head"));
        JsonObject rh = module.getAsJsonObject("roi_head");
        assertTrue(rh.has("bbox_roi_extractor"));
        assertTrue(rh.has("bbox_head"));
        assertTrue(rh.has("query_generator"));
        assertTrue(rh.has("pe"));

        if (rh.has("bbox_head")) {
            JsonObject bh = rh.getAsJsonObject("bbox_head");
            assertTrue(bh.has("transformer"));
            assertTrue(bh.has("bbox_coder"));
            int numClasses = bh.has("num_classes") ? bh.get("num_classes").getAsInt() : 1;
            assertTrue(numClasses > 0);
            if (bh.has("loss_cls")) {
                assertTrue(bh.get("loss_cls").isJsonObject());
            }
        }
        if (rh.has("pe")) {
            JsonObject pe = rh.getAsJsonObject("pe");
            assertTrue(pe.has("positional_encoding"));
        }
    }

    @Test
    public void testPublicTrainCfgDetections() throws Exception {
        for (String configPath : PUBLIC_EXP_CONFIGS) {
            JsonObject module = ConfigLoader.loadConfig(configPath).getAsJsonObject("module");
            if (module.has("train_cfg")) {
                JsonObject trainCfg = module.getAsJsonObject("train_cfg");
                if (trainCfg.has("detection_proposal")) {
                    JsonObject det = trainCfg.getAsJsonObject("detection_proposal");
                    assertTrue(det.has("score_thr"));
                    if (det.has("nms")) {
                        JsonObject nms = det.getAsJsonObject("nms");
                        assertTrue(nms.has("iou_threshold") || nms.has("class_agnostic"));
                    }
                }
            }
        }
    }

    @Test
    public void testPublicCodeWeightsVariations() throws Exception {
        String[] configPaths = {
            "configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep24.json",
            "configs/mv2d/exp/mv2d_r50_frcnn_two_frames_1408x512_ep72.json"
        };
        for (String configPath : configPaths) {
            JsonObject module = ConfigLoader.loadConfig(configPath).getAsJsonObject("module");
            JsonArray arr = module.getAsJsonObject("roi_head").getAsJsonObject("bbox_head")
                .getAsJsonArray("code_weights");
            assertTrue(arr.isJsonArray());
            assertEquals(10, arr.size());
            boolean allOnes = true;
            for (JsonElement e : arr) {
                assertTrue(e.isJsonPrimitive() && 
                    (e.getAsJsonPrimitive().isNumber() || e.getAsJsonPrimitive().isBoolean()));
                if (e.getAsDouble() != 1.0) allOnes = false;
            }
            assertFalse(allOnes, "code_weights should not all be 1.0 for public tests");
        }
    }

    @Test
    public void testPublicOperationOrderAndCp() throws Exception {
        for (String configPath : PUBLIC_EXP_CONFIGS) {
            JsonObject module = ConfigLoader.loadConfig(configPath).getAsJsonObject("module");
            JsonObject bboxHead = module.getAsJsonObject("roi_head").getAsJsonObject("bbox_head");
            if (!bboxHead.has("transformer")) continue;
            JsonObject tf = bboxHead.getAsJsonObject("transformer");
            if (!tf.has("decoder")) continue;
            JsonObject decoder = tf.getAsJsonObject("decoder");
            if (!decoder.has("transformerlayers")) continue;
            JsonObject tl = decoder.getAsJsonObject("transformerlayers");
            JsonElement oo = tl.get("operation_order");
            assertTrue(oo != null && (oo.isJsonArray()));
            JsonElement cp = tl.get("with_cp");
            assertTrue(cp.isJsonPrimitive() && (cp.getAsBoolean() == true || cp.getAsBoolean() == false));
        }
    }
}